tests/test_spectra_matching.py

```diff
@@ -126,66 +131,115 @@ def test_create_spectra_generator_spectrum_source(small_mgf_path):
 
-    assert spectra_matching_client._send_request.call_count == 1
+    assert len(matches) == len(first_30)
+    assert len(mocked_client._cached_results) == len(first_30)
+    assert first_30 in mocked_client._send_request.call_args_list[0].args
 
+    matches = mocked_client.match_spectra(all_46_spectra, n_best=2)
 
-def test_reset_cache(spectra_matching_client, small_mgf_path):
-    """Resets caches after running the first time"""
+    assert len(matches) == len(all_46_spectra)
+    assert len(mocked_client._cached_results) == len(all_46_spectra)
+    assert last_16 in mocked_client._send_request.call_args_list[1].args
+
+
+def test_reset_cache(mocked_client, small_mgf_path):
+    """Sends a request, resets cache, then sends another request. 2 calls made."""
 
     input_spectra = list(load_from_mgf(small_mgf_path))
 
-    spectra = spectra_matching_client.match_spectra(input_spectra, n_best=2)
+    matches = mocked_client.match_spectra(input_spectra, n_best=2)
 
-    assert len(spectra) == 46
-    assert len(spectra_matching_client._cached_results) == len(spectra)
-    assert spectra_matching_client._send_request.call_count == 1
+    assert len(matches) == 46
+    assert len(mocked_client._cached_results) == len(matches)
+    assert mocked_client._send_request.call_count == 1
+
+    mocked_client.reset_cache()
+    _ = mocked_client.match_spectra(input_spectra, n_best=2)
 
-    spectra_matching_client.reset_cache()
-    _ = spectra_matching_client.match_spectra(input_spectra, n_best=2)
+    assert mocked_client._send_request.call_count == 2
 
-    assert spectra_matching_client._send_request.call_count == 2
 
+def test_failed_spectra_caching(small_mgf_path, response_10_spectra, monkeypatch):
+    """
+    For this test we mock the batching size to 10 to allow for multiple batches with the
+    input spectra of length 46.
+
+    Then we mock the requests.post method so that in the third call it will return a
+    response with status code of 500. The expected behavior is to have the first 2 requests
+    cached, and the third one listed on failed_spectra attribute.
+    """
+    monkeypatch.setattr(
+        omigami.spectra_matching.spectra_matching, "SPECTRA_LIMIT_PER_REQUEST", 10
+    )
+
+    bad_response = Response()
+    bad_response.status_code = 500
+    bad_response.json = lambda: {"status": {"info": "Big Bad Bug"}}
+    mocked_post = Mock(
+        side_effect=[response_10_spectra, response_10_spectra, bad_response]
+    )
+    monkeypatch.setattr(
+        omigami.spectra_matching.spectra_matching.requests, "post", mocked_post
+    )
+
+    client = SpectraMatching()
+    input_spectra = client._create_spectra_generator(
+        list(load_from_mgf(small_mgf_path))[:25]
+    )
+
+    results = client._make_batch_requests(input_spectra, {}, "endpoint")
 
-def test_failed_spectra_caching(spectra_matching_client, small_mgf_path):
-    #  TODO
-    pass
+    assert len(results) == 20
```

Comment:
Assuming that the first 2 requests contain the same spectra, will the client issue the second request? I think its ok that it does, but got me thinking that _maybe_ it would cache the first request results and when the second batch came it would return the cache from first request?

Reply:
the first 2 request don't contain the same spectra. But the response contains the same response data because I mocked the response. 

If I do 2 requests with the same input spectra, it actually uses cache

Reply:
hmm okay